File: ling_chat/core/schemas/responses.py
# responses.py
from typing import Optional, Union

from pydantic import BaseModel

from .response_types import ResponseType


class BaseResponse(BaseModel):
    type: str
    duration: float = -1
    isFinal: bool = False


class ReplyResponse(BaseResponse):
    type: str = ResponseType.AI_REPLY
    character: Optional[str] = None
    roleId: Optional[int] = None
    emotion: str
    originalTag: str
    message: str
    ttsText: Optional[str] = None
    motionText: Optional[str] = None
    audioFile: Optional[str] = None
    originalMessage: str
    displayName: Optional[str] = None
    displaySubtitle: Optional[str] = None


class ChapterChangeResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_CHAPTER_CHANGE
    chapterName: str
    duration: float = 0


class ChoiceResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_CHOICE
    choices: list[str]
    allowFree: bool = False
    duration: float = 0


class ThinkingResponse(BaseResponse):
    type: str = ResponseType.AI_THINKING
    duration: float = 0
    isThinking: bool


class FreeDialogueResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_FREE_DIALGOUE
    switch: bool
    maxRounds: int
    endLine: str
    duration: float = 0


class ScriptBackgroundResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_BACKGROUND
    imagePath: str
    transition: float = 1


class ScriptPresentPicResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_PRESENT_PIC
    imagePath: str
    scale: float = 1
    duration: float = 0


class ScriptBackgroundEffectResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_BACKGROUND_EFFECT
    effect: str


class ScriptSoundResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_SOUND
    soundPath: str


class ScriptMusicResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_MUSIC
    musicPath: str


class ScriptNarrationResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_NARRATION
    displayName: Optional[str] = None
    text: str


class ScriptDialogResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_DIALOG
    emotion: str
    originalTag: str
    message: str
    motionText: Optional[str] = None
    audioFile: Optional[str] = None


class ScriptModifyCharacterResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_MODIFY_CHARACTER
    characterId: int
    emotion: Optional[str] = None
    action: Optional[str] = None


class ScriptPlayerResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_PLAYER
    text: str
    displayName: Optional[str] = None
    displaySubtitle: Optional[str] = None


class ScriptInputResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_INPUT
    hint: str


class ScriptEndResponse(BaseResponse):
    type: str = ResponseType.SCRIPT_END
    duration: float = 0
    isFinal: bool = True


# 所有响应类型
Response = Union[ReplyResponse, ScriptBackgroundResponse, ScriptNarrationResponse]

File: src-tauri/src/ai_service/game_system/script_engine/responses.rs
//! Serializable payload types for Tauri script events.
//!
//! Each struct matches the payload shape expected by the frontend event processors
//! in `src/core/events/processors/`. The frontend's `asEvent()` helper merges in
//! `type` and `duration` fields, so those are omitted here.

use serde::Serialize;

// ============================================================
// Tauri event name constants
// ============================================================

pub mod event_names {
    pub const SCRIPT_NARRATION: &str = "script:narration";
    pub const SCRIPT_PLAYER: &str = "script:player";
    pub const SCRIPT_CHAPTER_CHANGE: &str = "script:chapter-change";
    pub const SCRIPT_BACKGROUND: &str = "script:background";
    pub const SCRIPT_BACKGROUND_EFFECT: &str = "script:background-effect";
    pub const SCRIPT_MUSIC: &str = "script:music";
    pub const SCRIPT_SOUND: &str = "script:sound";
    pub const SCRIPT_PRESENT_PIC: &str = "script:present-pic";
    pub const SCRIPT_MODIFY_CHARACTER: &str = "script:modify-character";
    pub const SCRIPT_INPUT: &str = "script:input";
    pub const SCRIPT_CHOICE: &str = "script:choice";
    pub const SCRIPT_END: &str = "script:end";
    pub const SCRIPT_FREE_DIALOGUE: &str = "script:free-dialogue";
}

// ============================================================
// Payload types (fields match frontend `src/types/script.ts`)
// ============================================================

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NarrationPayload {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlayerPayload {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChapterChangePayload {
    pub chapter_name: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackgroundPayload {
    pub image_path: String,
    #[serde(default)]
    pub transition: f64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackgroundEffectPayload {
    pub effect: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MusicPayload {
    pub music_path: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SoundPayload {
    pub sound_path: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PresentPicPayload {
    pub image_path: String,
    #[serde(default = "default_scale")]
    pub scale: f64,
}

#[allow(dead_code)]
fn default_scale() -> f64 {
    1.0
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ModifyCharacterPayload {
    pub character_id: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emotion: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub action: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub clothes: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InputPayload {
    pub hint: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChoicePayload {
    pub choices: Vec<String>,
    #[serde(default)]
    #[serde(rename = "allowFree")]
    pub allow_free: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FreeDialoguePayload {
    #[serde(rename = "switch")]
    pub switch: bool,
    pub max_rounds: i32,
    pub end_line: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ScriptEndPayload {}
